web3walletcore/src/commonMain/kotlin/com/sonsofcrypto/web3walletcore/services/actions/ActionsService.kt

```diff
@@ -2,86 +2,64 @@ package com.sonsofcrypto.web3walletcore.services.actions
 
 import com.sonsofcrypto.web3lib.keyValueStore.KeyValueStore
 import com.sonsofcrypto.web3lib.services.networks.NetworksService
-import com.sonsofcrypto.web3walletcore.services.actions.Action.Type.*
+import com.sonsofcrypto.web3lib.utils.WeakRef
 
-interface ActionsServiceListener {
+interface ActionsListener {
     /** This will be called each time the list of Actions changes. */
     fun actionsUpdated()
 }
 
 interface ActionsService {
+    /** List of actions yet to be completed */
+    fun actions(): List<Action>
+    /** Mark action as completed */
+    fun markComplete(action: Action)
     /** Add a listener to the service */
-    fun addListener(listener: ActionsServiceListener)
+    fun addListener(listener: ActionsListener)
     /** Remove a listener from the service */
-    fun removeListener(listener: ActionsServiceListener)
-    /** Complete an action */
-    fun completeActionType(type: Action.Type)
-    /** List of actions yet to be completed */
-    fun outstandingActions(): List<Action>
+    fun removeListener(listener: ActionsListener)
 }
 
 class DefaultActionsService(
     private val store: KeyValueStore,
     private val networksService: NetworksService,
 ): ActionsService {
-    private var listeners = mutableListOf<ActionsServiceListener>()
-
-    override fun addListener(listener: ActionsServiceListener) {
-        listeners.add(listener)
-    }
-
-    override fun removeListener(listener: ActionsServiceListener) {
-        listeners.remove(listener)
-    }
 
-    override fun completeActionType(type: Action.Type) {
-        if (isActionCompleted(type)) return
-        store[type.identifier] = true
-        listeners.forEach { it.actionsUpdated() }
-    }
+    private var listeners: MutableList<WeakRef<ActionsListener>> = mutableListOf()
```

Comment:
does this needs to be a WeakRef? 🤔 You would register and unregister when you need too...not sure if is really needed WeakRef...

Reply:
As best practice it better if we go safer route. Cost us nothing. Avoids potentional memory issues. Just best practice.